### PR TITLE
PCHR-4367: Show Absence Modal for Entitlement Calculation For new Absence Periods

### PR DESCRIPTION
**Pull request by tunbola**
## Overview
When a new absence period is about to be created and the user clicks the save button, we need to show the user a modal to give an option of proceeding to the entitlements calculation page for the newly added period, the user may choose not to but the absence period will still get saved nonetheless whether user choses to proceed to entitlement calculation page or not.

## Before
- No modal prompt exists that allows a user to proceed to entitlement calculation page for new absence periods.

## After
- A modal is added that gives the user an option to proceed to entitlement calculation page for new absence period added.

scenario4
scenario3
scenario2
scenario1

## Technical Details
A confirmation modal for validating the absence period order on clicking the save button already exists. The new modal added in this ticket simply chains to that existing validation depending on the outcome to show the entitlement confirmation modal as seen in the first two Gifs.

```php
    AbsencePeriod.prototype._showEntitlementCalculationPrompt = function () {
      var confirmationMessage = 'The system will now update the staff members ' +
                                'leave entitlement';

      CRM.confirm({
        title: ts('Update Leave Entitlement?'),
        message: ts(confirmationMessage),
        width: '30%',
        options: {
          yes: ts('Proceed'),
          no: ts('Cancel')
        }
      })
        .on('crmConfirm:yes', this._processEntitlementConfirmation.bind(this))
        .on('crmConfirm:no', this._submitForm.bind(this));
    };
```

### CODE REVIEW

**Comment by igorpavlov** [comment]
@tunbola Wouldn't it be slightly better from the UX perspective to leave the save button "locked" (disabled) while redirecting to another page? It looks like any response to the prompt enables the button back but I think only "Cancel" that should disable it.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Form/AbsencePeriod.php [hunk 1]
```diff
@@ -137,6 +144,11 @@ private function addFields() {
       $this->getDAOFieldAttributes('weight'),
       true
     );
+
+    //This element will have its value set if the user
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Form/AbsencePeriod.php]:
```js
// This element is needed to store the state of the
// absence period entitlement calculation prompt
```
did I understand your comment correctly? if yes, would you mind enhancing it a bit like suggested?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/src/leave-absences/crm/app.form.absenceperiod.js [hunk 2]
```diff
@@ -44,6 +44,41 @@ define([
...
     };
 
+    /**
+     * Prompts the user via a modal to calculate entitlement for
+     * a new absence period
+     *
+     * @private
+     */
+    AbsencePeriod.prototype._showEntitlementCalculationPrompt = function () {
+      var confirmationMessage = 'The system will now update the staff members ' +
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/src/leave-absences/crm/app.form.absenceperiod.js]: No need to split, just put it on the next line:

```js
var confirmationMessage =
  'The system will now update the staff members leave entitlement';
```
or even as an exception:

```js
var confirmationMessage = 'The system will now update the staff members leave entitlement';
```

Please note, we are not **strictly** following the symbols amount rule, this is mostly for us to keep the code readable. This is just a long string, you will not gain any more info by reading it whole or just a half. So it's fine to keep it one-liner.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/src/leave-absences/crm/app.form.absenceperiod.js [hunk 3]
```diff
@@ -44,6 +44,41 @@ define([
...
+                                'leave entitlement';
+
+      CRM.confirm({
+        title: ts('Update Leave Entitlement?'),
+        message: ts(confirmationMessage),
+        width: '30%',
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/src/leave-absences/crm/app.form.absenceperiod.js]: Is this an OK practice to store width in JS and not CSS?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/src/leave-absences/crm/app.form.absenceperiod.js [hunk 4]
```diff
@@ -44,6 +44,41 @@ define([
...
+
+    /**
+     * Sets the value of the confirmEntitlement form element and
+     * submits the form.
+     *
+     * @private
+     */
+    AbsencePeriod.prototype._processEntitlementConfirmation = function () {
+      var confirmEntitlement = document.getElementsByName('confirmEntitlement')[0];
+      confirmEntitlement.value = 'yes';
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/src/leave-absences/crm/app.form.absenceperiod.js]: Please add an empty line above

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/src/leave-absences/crm/app.form.absenceperiod.js [hunk 5]
```diff
@@ -44,6 +44,41 @@ define([
...
+      this._submitForm();
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/src/leave-absences/crm/app.form.absenceperiod.js]: Please add an empty line above

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/src/leave-absences/crm/app.form.absenceperiod.js [hunk 6]
```diff
@@ -79,7 +114,26 @@ define([
...
+
+    /**
+     * Determines what action to take after the validate order Api
+     * callback, i.e what happens if the modal was not shown or if
+     * the confirmation modal was shown and the user clicks yes.
+     *
+     * @private
+     */
+    AbsencePeriod.prototype._processValidateOrder = function () {
+      var id = document.getElementsByName('_id')[0].value;
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/src/leave-absences/crm/app.form.absenceperiod.js]:
```js
var absencePeriodExists = !!document.getElementsByName('_id')[0].value;

if (absencePeriodExists) {
  this._submitForm();
} else {
  this._showEntitlementCalculationPrompt();
}
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/src/leave-absences/crm/app.form.absenceperiod.js (hunk 3 above)
Comment (tunbola) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/src/leave-absences/crm/app.form.absenceperiod.js]: Well, not sure, I added this to keep it consistent with the other modal in the file. Should I remove it? 
https://github.com/compucorp/civihr/blob/cc485f95e976a4b1f1309b14d688b9d023cd1016/uk.co.compucorp.civicrm.hrleaveandabsences/js/src/leave-absences/crm/app.form.absenceperiod.js#L156

**Pull request merged**